pkg/iac/scanners/terraformplan/snapshot/scanner_test.go

```diff
@@ -0,0 +1,130 @@
+package snapshot
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"testing"
+
+	"github.com/aquasecurity/trivy/pkg/iac/scan"
+	"github.com/aquasecurity/trivy/pkg/iac/scanners/options"
+	"github.com/samber/lo"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func initScanner(opts ...options.ScannerOption) *Scanner {
+	defaultOpts := []options.ScannerOption{
+		options.ScannerWithEmbeddedPolicies(false),
+		options.ScannerWithEmbeddedLibraries(true),
+		options.ScannerWithPolicyNamespaces("user"),
+		options.ScannerWithPolicyDirs("."),
+		options.ScannerWithRegoOnly(true),
+	}
+
+	opts = append(opts, defaultOpts...)
+	return New(opts...)
+}
+
+func TestScanner_Scan(t *testing.T) {
+	tests := []struct {
+		name        string
+		dir         string
+		expectedIDs []string
+	}{
+		{
+			name:        "one resource",
+			dir:         "just-resource",
+			expectedIDs: []string{"ID001"},
+		},
+		{
+			name:        "with local module",
+			dir:         "with-local-module",
+			expectedIDs: []string{"ID001"},
+		},
+		{
+			name:        "with remote module",
+			dir:         "with-remote-module",
+			expectedIDs: []string{"ID001"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			f, err := os.Open(filepath.Join("testdata", tt.dir, "tfplan"))
+			require.NoError(t, err)
+			defer f.Close()
+
+			policyFS := os.DirFS(filepath.Join("testdata", tt.dir, "checks"))
+
+			s := initScanner(options.ScannerWithPolicyFilesystem(policyFS))
+			result, err := s.Scan(context.TODO(), f)
+			require.NoError(t, err)
+
+			failed := result.GetFailed()
+
+			assert.Len(t, failed, len(tt.expectedIDs))
+
+			ids := lo.Map(failed, func(res scan.Result, _ int) string {
+				return res.Rule().AVDID
+			})
+			sort.Strings(ids)
+
+			assert.Equal(t, tt.expectedIDs, ids)
+		})
+	}
+}
+
+func Test_ScanFS(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		dir         string
+		expectedDir string
+	}{
+		{
+			dir:         "just-resource",
+			expectedDir: "main.tf",
+		},
+		{
+			dir:         "with-local-module",
+			expectedDir: path.Join("modules", "ec2", "main.tf"),
+		},
+		{
+			dir:         "with-remote-module",
+			expectedDir: path.Join("terraform-aws-modules", "s3-bucket", "aws", "main.tf"),
```

Comment:
Checking the path of a remote module is not reliable, as the result may depend on the order in which the tests are executed. If the module is not in the cache, the path is a reference to the git, otherwise the path to the module is in the cache.

Reply:
I wasn't able to make this test fail, is it flaky? Is there any way to test without the effects of the cache?

Reply:
1. clear the cache `rm -rf $TMPDIR/.aqua`.
2. comment out the first two test cases so that the test with the remote module is the first.
3. run test:
```bash
Error:      Not equal: 
            expected: "terraform-aws-modules/s3-bucket/aws/main.tf"
            actual  : "git::https:/github.com/terraform-aws-modules/terraform-aws-s3-bucket?ref=3a1c80b29fdf8fc682d2749456ec36ecbaf4ce14/main.tf"
```

Reply:
Since only the remote module is used in this test case, we only need to verify that the check detected misconfig. I'll update the test

Reply:
I created an issue in defsec asking to add the ability to skip cache to avoid fluky tests, but the issue is now closed. Can we move this issue to Trivy?

Reply:
> I created an issue in defsec asking to add the ability to skip cache to avoid fluky tests, but the issues are now closed. Can we move this issue to Trivy?

Yes please, if it's a small change you can just make a PR without the need to track with an issue.